File: apps/gateway/app/services/pricing.py
PRICING: dict[tuple[str, str], tuple[float, float]] = {
    ("anthropic", "claude-opus-4-7"):   (15.0,  75.0),
    ("anthropic", "claude-sonnet-4-6"): (3.0,   15.0),
    ("anthropic", "claude-haiku-4-5"):  (0.8,   4.0),
    ("openai",    "gpt-4o"):            (2.5,   10.0),
    ("openai",    "gpt-4o-mini"):       (0.15,  0.6),
    ("openai",    "o3"):                (10.0,  40.0),
    ("gemini",    "gemini-2.5-pro"):    (1.25,  10.0),
    ("gemini",    "gemini-2.5-flash"):  (0.075, 0.3),
    ("ollama",    "local"):             (0.0,   0.0),
}


def estimate_cost(provider: str, model: str, input_tokens: int, output_tokens: int) -> float:
    key = (provider, model)
    if key not in PRICING:
        key = next(
            (k for k in PRICING if k[0] == provider and model.startswith(k[1])),
            None,
        )
    if key is None:
        return 0.0
    input_price, output_price = PRICING[key]
    return (input_tokens / 1_000_000) * input_price + (output_tokens / 1_000_000) * output_price

File: packages/provider-adapters/src/pricing.ts
import type { ProviderPricing } from "@observaai/shared-types";

export const PRICING_TABLE: ProviderPricing[] = [
  { provider: "anthropic", model: "claude-opus-4-7",        inputPricePer1M: 15,   outputPricePer1M: 75   },
  { provider: "anthropic", model: "claude-sonnet-4-6",      inputPricePer1M: 3,    outputPricePer1M: 15   },
  { provider: "anthropic", model: "claude-haiku-4-5",       inputPricePer1M: 0.8,  outputPricePer1M: 4    },
  { provider: "openai",    model: "gpt-4o",                 inputPricePer1M: 2.5,  outputPricePer1M: 10   },
  { provider: "openai",    model: "gpt-4o-mini",            inputPricePer1M: 0.15, outputPricePer1M: 0.6  },
  { provider: "openai",    model: "o3",                     inputPricePer1M: 10,   outputPricePer1M: 40   },
  { provider: "gemini",    model: "gemini-2.5-pro",         inputPricePer1M: 1.25, outputPricePer1M: 10   },
  { provider: "gemini",    model: "gemini-2.5-flash",       inputPricePer1M: 0.075,outputPricePer1M: 0.3  },
  { provider: "ollama",    model: "local",                  inputPricePer1M: 0,    outputPricePer1M: 0    },
  { provider: "openrouter",model: "openrouter/auto",        inputPricePer1M: 0,    outputPricePer1M: 0    },
];

export function estimateCost(
  provider: string,
  model: string,
  inputTokens: number,
  outputTokens: number
): number {
  const pricing = PRICING_TABLE.find(
    (p) => p.provider === provider && (p.model === model || model.includes(p.model))
  );
  if (!pricing) return 0;
  return (
    (inputTokens / 1_000_000) * pricing.inputPricePer1M +
    (outputTokens / 1_000_000) * pricing.outputPricePer1M
  );
}
